x-pack/plugins/cases/server/routes/api/files/post_file.ts

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { schema } from '@kbn/config-schema';
+import { CASE_FILES_URL, MAX_FILE_SIZE } from '../../../../common/constants';
+import { createCaseError } from '../../../common/error';
+import { createCasesRoute } from '../create_cases_route';
+import type { attachmentApiV1 } from '../../../../common/types/api';
+import type { caseDomainV1 } from '../../../../common/types/domain';
+
+export const postFileRoute = createCasesRoute({
+  method: 'post',
+  path: CASE_FILES_URL,
+  params: {
+    params: schema.object({
+      case_id: schema.string(),
+    }),
+  },
+  routerOptions: {
+    access: 'public',
+    summary: 'Attach a file to a case',
+    tags: ['oas-tag:cases'],
+    body: {
+      /*
+       * The maximum file size allowed here is the upper limit allowed when registering the cases file
+       * kinds. We register the file kinds in x-pack/plugins/cases/public/files/index.ts.
+       *
+       * If the user configured a custom value the validation will still fail when calling the service.
+       */
+      maxBytes: MAX_FILE_SIZE,
+      accepts: 'multipart/form-data',
+    },
+  },
+  handler: async ({ context, request, response }) => {
+    try {
+      const caseContext = await context.cases;
+      const casesClient = await caseContext.getCasesClient();
+
+      const fileRequest = request.body as attachmentApiV1.PostFileAttachmentRequest;
```

Comment:
I noticed that the following information is provided by `hapi` about the file in `request.body.file`. Example:

```
{
  filename: 'foo.json',
  headers: {
    'content-disposition': 'form-data; name="file"; filename="foo.json"',
    'content-type': 'application/json'
  }
}
```

Then you can get all the information you want as:

```
import mime from 'mime-types';

const fileExtension = extname(file.hapi.filename).toLowerCase();
const mimeType = mime.lookup(fileExtension);
```

I am wondering if we can get the information like this do we need the `fileName` and the `mimeType` in the body? It will reduce users extra steps to do and figure out. Wdyt?

Reply:
> I noticed that the following information is provided by hapi about the file in request.body.file.

Do you mean `request.body.file.hapi`?

I see it too

```
  hapi: {
    filename: 'requirements.txt',
    headers: {
      'content-disposition': 'form-data; name="file"; filename="requirements.txt"',
      'content-type': 'text/plain'
    }
  },
```

Is there maybe a cleaner way to access this? So now I wonder if this would allow us not to use `multipart/form-data`.

Reply:
Nevermind, I saw how fleet does it(I assume that is where you saw it too 😅 ) and they access it directly. `fileStream.hapi.filename`.

Also, I tried with `content-type octet-stream` for a bit and keep running into 415s. I think I will keep the multipart form and make filename and mime type optional. Then The user can pick if he wants to set them or if he wants the values from the file metadata(that we get with hapi).

Reply:
Yes, `request.body.file.hapi`! My inspiration is mostly from here `packages/core/saved-objects/core-saved-objects-server-internal/src/routes/import.ts` and here `src/plugins/files/server/routes/file_kind/upload.ts`. Yes, let's stick with the multipart form. As we discuss offline, better to allow only the filename.